### PR TITLE
MGUsup

### PR DESCRIPTION
**Pull request by aematts**
## PR Summary

Included code and documentation for a Mie-Gruneisen EOS based on a linear Us-up relation.
It is a full EOS with consistent temperature.

## PR Checklist

- [ x] Adds a test for any bugs fixed. Adds tests for new features.
- [ x] Format your changes by using the `make format` command after configuring with `cmake`.
- [ x] Document any new features, update documentation for changes made.
- [x] Make sure the copyright notice on any files you modified is up to date.
- [x] After creating a pull request, note it in the CHANGELOG.md file.
- [x] LANL employees: make sure tests pass both on the github CI and on the Darwin CI

If preparing for a new release, in addition please check the following:
- [ ] Update the version in cmake.
- [ ] Move the changes in the CHANGELOG.md file under a new header for the new release, and reset the categories.

### CODE REVIEW

File: doc/sphinx/src/models.rst
Comment (jhp-lanl) [inline doc/sphinx/src/models.rst:783]:
```suggestion
we now extrapolate from the temperature on the Hugoniot, :math:`T_H(\rho)`, and not 
```

Looks like this isn't rendering properly. I _think_ this change fixes it

File: doc/sphinx/src/models.rst
Comment (jhp-lanl) [inline doc/sphinx/src/models.rst:787]:
```suggestion
With this consistent temperature we can derive an entropy in a similar way as for the Vinet EOS,

.. math::
```
This isn't displaying correctly either. I think it's just missing a carriage return.

File: doc/sphinx/src/models.rst
Comment (jhp-lanl) [inline doc/sphinx/src/models.rst:793]:
```suggestion
where :math:`\eta` is a measure of compression given by

.. math::
```

File: doc/sphinx/src/models.rst
Comment (jhp-lanl) [inline doc/sphinx/src/models.rst:803]:
```suggestion
shock jump conditions assuming a linear :math:`U_s`-:math:`u_p` relation,

.. math::
```

File: doc/sphinx/src/models.rst
Comment (jhp-lanl) [inline doc/sphinx/src/models.rst:797]:
```suggestion
This is convenient because :math:`\eta = 0` when :math:`\rho = \rho_0`,
```

File: doc/sphinx/src/models.rst
Comment (jhp-lanl) [inline doc/sphinx/src/models.rst:828]:
```suggestion
it is

.. math::
```

**Review by jhp-lanl — APPROVED** [review]
Great documentation @aematts ! You're also welcome to put a blurb in the other Gruniesen EOS to encourage the user to use the thermodynamically consistent one if they are able. It's up to you though whether you want to add this.

As long as things compile, this looks good to me!

**Comment by Yurlungur** [comment]
@aematts as this is marked WIP do you want review now? Or do you want us to wait? I see @jhp-lanl already reviewed. :)

**Comment by aematts** [comment]
> @aematts as this is marked WIP do you want review now? Or do you want us to wait? I see @jhp-lanl already reviewed. :)

I want reviews but not merge just yet. I am fixing the documentation since I am only using the pipeline for compiling it. So I need to do some commits more. I will remove the WIP when I am ready for merge.

**Comment by aematts** [comment]
And both of you: THANKS for getting on it so fast.

**Comment by Yurlungur** [comment]
> > @aematts as this is marked WIP do you want review now? Or do you want us to wait? I see @jhp-lanl already reviewed. :)
> 
> I want reviews but not merge just yet. I am fixing the documentation since I am only using the pipeline for compiling it. So I need to do some commits more. I will remove the WIP when I am ready for merge.

:+1: Understood. I will make sure to review this by the end of today.

File: singularity-eos/eos/eos_mgusup.hpp
Comment (jhp-lanl) [inline singularity-eos/eos/eos_mgusup.hpp:153]: Would it make sense to enforce a maximum compressibility here instead?

File: singularity-eos/eos/eos_mgusup.hpp
Comment (jhp-lanl) [inline singularity-eos/eos/eos_mgusup.hpp:233]: Similarly to the above comment, I think this would lead to a negative bulk modulus when the maximum compression limit is exceeded.

**Review by jhp-lanl — CHANGES_REQUESTED** [review]
I think it would be good to compute a maximum density for this EOS to ensure that sound speeds aren't imaginary.

You can see we used this approach in the other Gruneisen EOS, but with a much more complicated procedure for calculating the maximum density (since it's a cubic Us-up fit).

I would recommend mirroring this approach where the user is allowed to provide a maximum density or not. If the maximum density is used, then we set it and perhaps warn the user if it's greater than the singularity in the denominator of the Hugoniot expression. If the user does not provide a maximum density, then we provide one based off of the Hugoniot fit parameter.

See this section of the other Gruneisen EOS for an example of how to do this.
https://github.com/lanl/singularity-eos/blob/f7df74f90d53e97d048cd230eec1d6203d7c37d2/singularity-eos/eos/eos_gruneisen.hpp#L40

**Comment by aematts** [comment]
I have two more questions:
1) where do I find the correct copyright text for updating the copyright notice?
2) I do not know how to run on the GPUs on Darwin so can someone else make sure this run correctly on Darwin?
(There are a few failed tests that I have not touched at all and they failed already before I started working on the files. What do I do with these?)

**Comment by jhp-lanl** [comment]
>     1. where do I find the correct copyright text for updating the copyright notice?

Just copy the old one and update the date to 2024. Just worry about the files you change. We'll do an automated update later.

>     2. I do not know how to run on the GPUs on Darwin so can someone else make sure this run correctly on Darwin?
>        (There are a few failed tests that I have not touched at all and they failed already before I started working on the files. What do I do with these?)

It's easiest to use the gitlab CI. I'll post some instructions in mattermost for how to do this.

File: singularity-eos/eos/eos_mgusup.hpp
Comment (aematts) [reply singularity-eos/eos/eos_mgusup.hpp:153]: So, I think I warn and throw the code if 1-s \eta gets negative. I think I even point out that the s and \eta combined is non-physical.

**Comment by Yurlungur** [comment]
> I have two more questions:
> 
>     1. where do I find the correct copyright text for updating the copyright notice?

All this means is that in each file you touch, make sure that there's a blurb like the one below. And that the date includes the current calendar year as the end date... e.g., 2024.
```
//------------------------------------------------------------------------------
// © 2021-2024. Triad National Security, LLC. All rights reserved.  This
// program was produced under U.S. Government contract 89233218CNA000001
// for Los Alamos National Laboratory (LANL), which is operated by Triad
// National Security, LLC for the U.S.  Department of Energy/National
// Nuclear Security Administration. All rights in the program are
// reserved by Triad National Security, LLC, and the U.S. Department of
// Energy/National Nuclear Security Administration. The Government is
// granted for itself and others acting on its behalf a nonexclusive,
// paid-up, irrevocable worldwide license in this material to reproduce,
// prepare derivative works, distribute copies to the public, perform
// publicly and display publicly, and to permit others to do so.
//------------------------------------------------------------------------------
```

>     2. I do not know how to run on the GPUs on Darwin so can someone else make sure this run correctly on Darwin?

Yeah I will trigger the tests. I can also walk you through how to do it at some point. The way to do it is to push your branch to the mirror on re-git and create a pull request there. We never merge that PR but it will trigger the Darwin tests and report their status. I did it for this branch here:
https://re-git.lanl.gov/xcap/oss/singularity-eos/-/merge_requests/92

>        (There are a few failed tests that I have not touched at all and they failed already before I started working on the files. What do I do with these?)

I don't see any failed tests at least reported on github? Were they in the full test suite? We'll see what it says on Darwin.

File: singularity-eos/eos/eos_mgusup.hpp
Comment (aematts) [reply singularity-eos/eos/eos_mgusup.hpp:153]: I have put in this comment where it will break the code, that is, in HugTemperatureFromDensity. It would definitely break TBulkModulusFromDensityTemperature but not necessarily HugPressureFromDensity.
Let me think about what is the correct thing to do here, from physics and thermodynamics.

Comment (jhp-lanl) [reply singularity-eos/eos/eos_mgusup.hpp:153]: > So, I think I warn and throw the code if 1-s \eta gets negative. I think I even point out that the s and \eta combined is non-physical.

Ah I see you thrown in the `HugTemperatureFromDensity()` function, but unless I'm misunderstanding the code flow, I don't think this would be triggered for e.g. `PressureFromDensityInternalEnergy()`.

> Let me think about what is the correct thing to do here, from physics and thermodynamics.

Sounds good 👍 .

Our solution in the other Gruneisen EOS where we just impose a maximum was mainly an attempt to mirror Pagosa's implementation. However, we gave thought to other approaches like returning the infinity as the appropriate limit, but we decided against approaches like that since they were more tricky from a numerical procedure. With our approach, we just selected a maximum density just before the singularity in the hope that the condition would be self-correcting (i.e. such a high pressure would immediately cause the material to resist further compression).

Regardless though, our approach isn't perfect and may benefit from improvement. I'll look forward to your future thoughts.

File: doc/sphinx/src/models.rst
Comment (Yurlungur) [inline doc/sphinx/src/models.rst:762]:
```suggestion
This version implements the exact thermodynamic temperature on the Hugoniot and also adds an entropy.
```

File: singularity-eos/eos/eos_mgusup.hpp
Comment (jhp-lanl) [reply singularity-eos/eos/eos_mgusup.hpp:153]: @jonahm-LANL feel free to correct my memory of the prior density limit discussion (from two years ago?)

File: test/test_eos_mgusup.cpp
Comment (Yurlungur) [inline test/test_eos_mgusup.cpp:525]: :+1: nice to see this test.

File: test/test_eos_mgusup.cpp
Comment (Yurlungur) [inline test/test_eos_mgusup.cpp:1]: I only skimmed this file, but it looks good. These tests look like they run on GPU and fairly complete.

File: singularity-eos/eos/eos_mgusup.hpp
Comment (Yurlungur) [inline singularity-eos/eos/eos_mgusup.hpp:126]: It's good practice to name private functions with a leading or trailing underscore. Can you add one to this name?

File: singularity-eos/eos/eos_mgusup.hpp
Comment (Yurlungur) [reply singularity-eos/eos/eos_mgusup.hpp:153]: It also doesn't have to be consistent between this one and the Pagosa-based one. So long as we document what we do I think any reasonable choice is fine.

File: singularity-eos/eos/eos_mgusup.hpp
Comment (Yurlungur) [inline singularity-eos/eos/eos_mgusup.hpp:167]: Hmm. Related to the discussion with @jhp-lanl  above, I would prefer if there was a way to enforce that the code return something reasonable here, even with bad input... This can be a discussion for later, though, when people are actually using the model. 

I just worry that we are likely to hit this bound in real simulations and we don't want somebody's run to crash because of this. Maybe we just bound f1 > 0?

**Review by Yurlungur — CHANGES_REQUESTED** [review]
Thanks for this, @aematts overall, this looks great! I have a few small nitpicks, but from my perspective this is basically good to go.

Whatever we decide to do about the compression limit, we should document. It's fine for things to fail in compression at runtime for now, but we should warn in the documentation. And we should think about a more robust solution longer term. For my money I would prefer if the code failed at initialization or not at all, as this would help deal with the garbage inputs that we are likely to receive from host codes.

File: singularity-eos/eos/eos_mgusup.hpp
Comment (Yurlungur) [reply singularity-eos/eos/eos_mgusup.hpp:153]: > @jonahm-LANL feel free to correct my memory of the prior density limit discussion (from two years ago?)

I think you remember correctly.

File: test/test_eos_mgusup.cpp
Comment (aematts) [reply test/test_eos_mgusup.cpp:525]: Oops, I have the wrong variable names ;). Those are for Vinet. They have the correct values (I did a halfway check but forgot to go all the way) but I will change them to what I use for MGUsup.

File: singularity-eos/eos/eos_mgusup.hpp
Comment (aematts) [reply singularity-eos/eos/eos_mgusup.hpp:126]: OK. Sounds good.

File: singularity-eos/eos/eos_mgusup.hpp
Comment (aematts) [reply singularity-eos/eos/eos_mgusup.hpp:167]: We need to discuss this. It cannot be tested at initialization since \rho is varying, and thus \eta and thus s*\eta. A maximum \eta can be calculated of course (\eta_max = 1/s) and with that max Hugoniot pressure and max Hugoniot temperature. If this happens they should really not use these EOSs but use the PowerMG that I will implement next.

Comment (jhp-lanl) [reply singularity-eos/eos/eos_mgusup.hpp:167]: @aematts I'm happy to discuss this here or Mattermost but if you'd like to meet, I'm happy to do that too. Just let me know

**Comment by Yurlungur** [comment]
@aematts let us know when you'd like a re-review, so we don't miss it.

File: singularity-eos/eos/eos_mgusup.hpp
Comment (aematts) [reply singularity-eos/eos/eos_mgusup.hpp:153]: I resolve this but continue to think about it and discuss on Mattermost

File: doc/sphinx/src/models.rst
Comment (aematts) [reply doc/sphinx/src/models.rst:762]: Fixed

File: test/test_eos_mgusup.cpp
Comment (aematts) [reply test/test_eos_mgusup.cpp]: Fixed

File: singularity-eos/eos/eos_mgusup.hpp
Comment (aematts) [reply singularity-eos/eos/eos_mgusup.hpp]: Fixed

File: singularity-eos/eos/eos_mgusup.hpp
Comment (aematts) [reply singularity-eos/eos/eos_mgusup.hpp:167]: Continued on Mattermost

File: singularity-eos/eos/eos_mgusup.hpp
Comment (aematts) [reply singularity-eos/eos/eos_mgusup.hpp:233]: This will throw the code since it uses HugTemperatureFromDensity where this would be caught. We leave the discussion on how to do this properly to later after resolving the issue via Mattermost and/or other channels.

**Comment by aematts** [comment]
@jhp-lanl and @Yurlungur: The PR is now ready for final review.

**Review by Yurlungur — APPROVED** [review]
We can leave the discussion about what to do in compression for later. I am willing to approve and merge this now, pending @jhp-lanl 's approval.

Thanks for you work on this @aematts !

**Comment by aematts** [comment]
Can you NOT delete my branch, please. I would like to use it again for the PowerMG.

**Comment by Yurlungur** [comment]
> Can you NOT delete my branch, please. I would like to use it again for the PowerMG.

I changed the repository settings, the branch shouldn't be automatically deleted now.

**Comment by jhp-lanl** [comment]
> > Can you NOT delete my branch, please. I would like to use it again for the PowerMG.
> 
> I changed the repository settings, the branch shouldn't be automatically deleted now.

Just to be clear here though, the branch is only deleted from the _remote_ repository on github/re-git. The branch remains in your local checkout, so you can continue working on the branch even after it has been merged. Whenever you push the branch again, it will then exist in the remote repo as well.

**Review by jhp-lanl — APPROVED** [review]
I created an issue about the maximum compressibility issue (see https://github.com/lanl/singularity-eos/issues/332) so we can address it in the future. I'll approve the code as-is

**Comment by aematts** [comment]
Thank you guys! @jhp-lanl , that was exactly how I did it last time but I just did not want to do it again since I will go on with PowerMG immediately.

**Pull request merged**
